File: example/fs/Samples/GroupTest.fs
namespace Samples

open AUTD3Sharp
open AUTD3Sharp.Gain
open AUTD3Sharp.Modulation
open AUTD3Sharp.Utils
open type AUTD3Sharp.Units

module GroupByDeviceTest =
    let Test<'T> (autd : Controller<'T>) = 
        (Silencer.Default()) |> autd.Send;

        autd.Group(fun dev -> match dev.Idx with
                                | 0 ->  "null"
                                | 1 ->  "focus"
                                | _ ->  null
             )
            .Set("null", new Static(), new Null())
            .Set("focus", new Sine(150u * Hz), new Focus(autd.Geometry.Center + new Vector3d(0,0,150)))
            .Send();

module GroupByTransducerTest =
    let Test<'T> (autd : Controller<'T>) = 
        (Silencer.Default()) |> autd.Send;

        let cx = autd.Geometry.Center.X;
        let g1 = new Focus(autd.Geometry.Center + Vector3d(0., 0., 150.));
        let g2 = new Null();

        let grouping (dev: Device) (tr: Transducer) =
            if (tr.Position.X < cx) then "focus" :> obj else "null" :> obj
        let g = (new Group(grouping)).Set("focus", g1).Set("null", g2);
        let m = new Sine(150u * Hz);

        (m, g) |> autd.Send;

File: example/cs/Samples/GroupTest.cs
using AUTD3Sharp;
using AUTD3Sharp.Gain;
using AUTD3Sharp.Modulation;
using AUTD3Sharp.Utils;
using static AUTD3Sharp.Units;

namespace Samples;

internal static class GroupByDeviceTest
{
    public static void Test<T>(Controller<T> autd)
    {
        var config = Silencer.Default();
        autd.Send(config);

        autd.Group(dev =>
           {
               return dev.Idx switch
               {
                   0 => "null",
                   1 => "focus",
                   _ => null
               };
           })
           .Set("null", (new Static(), new Null()))
           .Set("focus", (new Sine(150 * Hz), new Focus(autd.Geometry.Center + new Vector3d(0, 0, 150))))
           .Send();
    }
}


internal static class GroupByTransducerTest
{
    public static void Test<T>(Controller<T> autd)
    {
        var config = Silencer.Default();
        autd.Send(config);

        var cx = autd.Geometry.Center.X;
        var g1 = new Focus(autd.Geometry.Center + new Vector3d(0, 0, 150));
        var g2 = new Null();

        var g = new Group(
            (_, tr) => tr.Position.X < cx ? "focus" : "null"
            ).Set("focus", g1).Set("null", g2);

        var m = new Sine(150 * Hz);

        autd.Send(m, g);
    }
}
